scripts/research/lib/research_harness.py

```
@@ -0,0 +1,579 @@
+@dataclass(frozen=True)
+class OutcomeAccessPolicy:
+    stage: str
+    start_inclusive_ms: int
+    end_exclusive_ms: int
+    allowed_years: tuple[int, ...]
+
+    def __post_init__(self) -> None:
+        if self.stage != DISCOVERY_STAGE:
+            raise ValueError(
+                f"v1 authorizes only {DISCOVERY_STAGE!r}, got {self.stage!r}"
+            )
```

Ruff (0.16.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)